entropy: add a config variable to include in image

antonlacon opened the pull request:
Entropy is a custom solution for getting initial randomness loaded into /dev/random, based on loading information at bootup from pseudorandomness generated at the last shutdown.

This moves entropy from being a dependency of systemd to virtual/image. It also moves its inclusion in the image behind the new variable, `HARDWARE_RNG`. Setting to 'yes' does not include entropy. This sets it to 'no' in distribution options.

Since kernel 3.16, if a hardware rng is available, it will contribute randomness to /dev/random (https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=be4000bc4644d027c519b6361f5ae3bbfc52c347). I'm fuzzy on how it interacts with CONFIG_TRUST_RANDOM_CPU, but with that not set, I'm still achieving ~30MB/s dd output from /dev/random, so I'm pretty sure it's still contributing.

RPi's all have a hwrng, so don't include entropy. Tested w/ RPi3.

chewitt reviewed (approved):
LGTM .. ping @CvH for tagging the release to pull it into

mglae commented:
A few thoughts (RFC):

1.) Please rename the option to USE_ENTROPY or similar because is does not change anything to HW RNG. You already did it in the commit description `distro: add toggle to include entropy package`.

2.) Why not only disabling the service by default to be prepared for failing HW RNGs on some CPUs (when not looking on RPIs)

2a) To make 2.) more easy we can simplify the entropy package. I already changed it for [9.2](http://ix.io/3rPl)  (because I did use the code to speed up the boot time of Linux VMs in the past) and don't expect issues after porting to LE10.

antonlacon commented:
1). Renamed to USE_ENTROPY.
2a). Added the cleanup commit. Looks ok to me.

2). I added two checks to the scripts:

On startup, it checks `/proc/sys/kernel/random/entropy_avail` to be above a trigger threshold. This is measured in bits, and I arbitrarily picked 128. If it is below the threshold, a message is written to the journal. On my RPi3, this script executes <2 seconds after systemd enters startup, and it's above this threshold.

On shutdown, just refresh the data if `random.data` already exists. The startup portion already determined if it was necessary.

I looked at an alternative approach of whether a hwrng registered itself (`/sys/class/misc/hw_random/current`). On systems that trust the cpu's rng instructions (x86's rdrand), it doesn't register a hwrng, but will have plenty of entropy available.

Added license/copyright information. Git shows @lrusak as initial author, so please comment if you want a change.

lrusak commented:
Is there any real benefit to making this optional? Even with hw rng this may help out in some situations.

I believe the original reason it was added was because there wasn't enough random entropy so startup was delayed and some wireless connections couldn't be established. This should probably be checked.

Making some features only available on certain platforms is a step backwards in my opinion. We should really be trying to make sure that the builds are as similar as possible (they will be different no matter what I know).

The license update is fine by me though.

vpeter4 commented on packages/sysutils/entropy/sources/add-entropy:
```diff
@@ -1,15 +1,33 @@
+
+# if available entropy does not need a kickstart, exit
+if os.path.isfile("/proc/sys/kernel/random/entropy_avail"):
+  available_entropy=subprocess.check_output("cat /proc/sys/kernel/random/entropy_avail", shell=True, text=True)
```
Use python file read instead of subprocess?

mglae commented:
> Is there any real benefit to making this optional? Even with hw rng this may help out in some situations.

Agree. Furthermore the additional entropy is mixed into the state and not replacing anything.
 
> I believe the original reason it was added was because there wasn't enough random entropy so startup was delayed and some wireless connections couldn't be established. This should probably be checked.

Yes, the wireless connection issues caused the creation of the package.

In the meantime there is much improvement in the RNG kernel code. But when booting LE inside a VM as the worst environment regarding to entropy the CRNG init is still delayed until the entropy is added:

```
Jul 04 15:51:00.559648 192.168.11.138 systemd[1]: Starting Add random entropy from file...
[...]
Jul 04 15:51:00.610533 192.168.11.138 kernel: random: crng init done
Jul 04 15:51:00.623144 192.168.11.138 systemd[1]: Finished Add random entropy from file.
```

> Making some features only available on certain platforms is a step backwards in my opinion. We should really be trying to make sure that the builds are as similar as possible (they will be different no matter what I know).

... and keep the code as simple as possible avoiding hard to remind conditions.

CvH commented:
gtg ?

mglae commented:
We should not rely on "magic" numbers of current CRNG implementation. Future CRNG changes may result in unexpected behavior.

E.g. there is an replacement at http://chronox.de/lrng.html

antonlacon commented:
> gtg ?

No, I think this one's done.
